### ISSUE TITLE
Training pipeline?

### ISSUE THREAD

**Issue by opendeeple** [isantesteban/snug#11]
Hi, on training, I have deformed and skinned garment from snug model and I fixed collision as post processing on the predicted garment. Then I am using loss functions to improve learning.
So, I want to ask, am I used right pipeline or need some changes?
Example code:
```py
# call snug model
# the snug model return predicted ang skinned garment
v_garment, hidden_states = snug(
  [shapes, poses, trans_vel],
  hidden_states=hidden_states
  trans=trans
)
# nearest neighbour indices
v_indices = find_index_sequence(
  vc=v_garment,
  vb=v_body
)
# fix collision | postprocessing
v_garment = fix_collision_sequence(
  vc=v_garment,
  vb=v_body,
  nb=v_normals, # body vertex normals
  v_indices=v_indices # nearest neighbour indices
)
# inertia loss
inertia = inertial_term_sequence(
  x=v_garment,
  mass=snug.cloth.v_mass
)
# reshape features for next losses
v_garment = tf.reshape(tensor=v_garment, shape=[-1, *v_garment.shape[2:]])
v_body    = tf.reshape(tensor=v_body,    shape=[-1, *v_body.shape[2:]])
v_normals = tf.reshape(tensor=v_normals, shape=[-1, *v_normals.shape[2:]]) # body normal vertices
v_indices = tf.reshape(tensor=v_indices, shape=[-1, v_indices.shape[2]]) # nearest neighbour indices
# collision loss
penalty_collision = collision_penalty(
  va=v_garment,
  vb=v_body,
  nb=v_normals,
  v_indices=v_indices
)
# stretch loss
energy_stretch = stretching_energy(
  v=v_garment,
  cloth=snug.cloth
)
# bending loss
energy_bend = bending_energy(
  v=v_garment,
  cloth=snug.cloth
)
# gravity loss
energy_gravity = gravitational_energy(
  x=v_garment,
  mass=snug.cloth.v_mass
)
```
Thanks for your attention!

**Comment by hsyntemiz** [comment]
Hi, 
As you do, I also try to replicate the training part. I have a similar pipeline. But, I couldn't succeed up to now. Do we need postprocessing in training? I am not sure.

I also have some additional gradient issues. If I don't multiply energy_stretch with a small value, loss terms become NaN immediately. But, these are energy terms. I think it is not correct to add coefficients for energy terms.  

loss_value = energy_bend + g_energy + inertia + 0.01*energy_stretch + penalty_collision 

I am open to collaborate to replicate training part. 


- One sample line from training log for energy values. Strech energy is commonly more than others (~1000x or much more)
Bending Energy:0.0080470,Gravitation: 0.058480,Inertia:8.7634e-05, Stretch: 1.305195,Penalty: 9.94163e-05

**Comment by isantesteban** [comment]
Hi,

We don't use the postprocess during training, only the collision penalty. 

Here are the values of the energies during the training process, we train for 10 epochs using a learning rate of 0.001 and then we resume the training for 100 additional epochs using a learning rate of 0.0001:

Untrained: 
> **full loss:** 392.6,  **bending**: 1.273, **collisions**: 8.24e-4, **gravity**: 0.053, **inertia**: 0.046, **stretch**: 391

After 10 epochs (LR 0.001):
> **full loss:** 0.782,  **bending**: 0.025, **collisions**: 1.8e-4, **gravity**: 0.054, **inertia**: 0.043, **stretch**: 0.62

After 100 more epochs (LR 0.0001):
> **full loss:** 0.256,  **bending**: 0.011, **collisions**: 9.1e-5, **gravity**: 0.033, **inertia**: 0.031, **stretch**: 0.16

Hope this helps!

**Comment by fishfishson** [comment]
Hi @isantesteban, thx for your great work and codes!
In your paper, the model only requires 3 frames for training. Does it manes any consecutive 3 frames of all sequences?
BTW, do you plan to share data preprocessing codes later?
Looking forward to your reply!

**Comment by liren2515** [comment]
> Hi,
> 
> We don't use the postprocess during training, only the collision penalty.
> 
> Here are the values of the energies during the training process, we train for 10 epochs using a learning rate of 0.001 and then we resume the training for 100 additional epochs using a learning rate of 0.0001:
> 
> Untrained:
> 
> > **full loss:** 392.6,  **bending**: 1.273, **collisions**: 8.24e-4, **gravity**: 0.053, **inertia**: 0.046, **stretch**: 391
> 
> After 10 epochs (LR 0.001):
> 
> > **full loss:** 0.782,  **bending**: 0.025, **collisions**: 1.8e-4, **gravity**: 0.054, **inertia**: 0.043, **stretch**: 0.62
> 
> After 100 more epochs (LR 0.0001):
> 
> > **full loss:** 0.256,  **bending**: 0.011, **collisions**: 9.1e-5, **gravity**: 0.033, **inertia**: 0.031, **stretch**: 0.16
> 
> Hope this helps!

Hello,

Are these values averaged by the number of vertices or just the sum of all vertices or faces or edges?

**Comment by liren2515** [comment]
@hsyntemiz Hello,

Have you succeeded in replicating the training. I'm trying to replicate with pytorch, but the loss seems too large. The untrained loss is like:
strain:95793.54688,  bending:22.87846,  gravity:27.13613,  inertia:11.21035,  collision:18.52707

Do you have the same observation?

**Comment by isantesteban** [comment]
> Are these values averaged by the number of vertices or just the sum of all vertices or faces or edges?

These are the values returned by the loss functions included in this repository, which do not average by the number of vertices/faces/edges

> In your paper, the model only requires 3 frames for training. Does it manes any consecutive 3 frames of all sequences?

We use the pose sequences from [this repository](https://github.com/isantesteban/vto-dataset) and divide each sequence in subsequences of 3 frames

> BTW, do you plan to share data preprocessing codes later?

We are not allowed to share additional code, sorry for the inconvenience

**Comment by liren2515** [comment]
@isantesteban Thanks for your information! Do you use some specific initialization strategy for the network in the training?

**Comment by opendeeple** [comment]
> Hi,
> 
> We don't use the postprocess during training, only the collision penalty.
> 
> Here are the values of the energies during the training process, we train for 10 epochs using a learning rate of 0.001 and then we resume the training for 100 additional epochs using a learning rate of 0.0001:
> 
> Untrained:
> 
> > **full loss:** 392.6,  **bending**: 1.273, **collisions**: 8.24e-4, **gravity**: 0.053, **inertia**: 0.046, **stretch**: 391
> 
> After 10 epochs (LR 0.001):
> 
> > **full loss:** 0.782,  **bending**: 0.025, **collisions**: 1.8e-4, **gravity**: 0.054, **inertia**: 0.043, **stretch**: 0.62
> 
> After 100 more epochs (LR 0.0001):
> 
> > **full loss:** 0.256,  **bending**: 0.011, **collisions**: 9.1e-5, **gravity**: 0.033, **inertia**: 0.031, **stretch**: 0.16
> 
> Hope this helps!

After 10 epochs (LR 0.001)

**full loss**: 1.22266, **inertia**: 0.01243, **stretch**: 1.04558, **bending**: 0.03306, **gravity**: 0.04731, **collisions**: 0.08428

After 100 epochs (LR 0.0001)

**full loss**: 0.20482, **inertia**: 0.01269, **stretch**: 0.12848, **bending**: 0.01619, **gravity**: 0.03859, **collisions**: 0.00887

Result:
![text2photo-fae1a4c9990dc3ba4a1d0823b6ce2814](https://user-images.githubusercontent.com/74002741/172907730-026104d7-b73b-4b2e-8c5b-cb4a4272c56b.jpeg)

Loss functions is looking as correct but our result is not perfect, so can you help me for improve it with your offers & analysis?
Thanks for your attention!!!

**Comment by hsyntemiz** [comment]
@opendeeple Did you apply post-processing to final meshes?

**Comment by opendeeple** [comment]
@hsyntemiz Yes to final meshes but on training I did not use post processing.

**Comment by hsyntemiz** [comment]
@opendeeple Do you plan to publish your training implementation? "Unofficial implementation" can attract many people. 
The authors help community too much, but they have obvious limitation about sharing training codes.

**Comment by hsyntemiz** [comment]
@liren2515 Not yet. 
> Do you have the same observation?

**Comment by opendeeple** [comment]
> @opendeeple Do you plan to publish your training implementation? "Unofficial implementation" can attract many people. The authors help community too much, but they have obvious limitation about sharing training codes.

@hsyntemiz Sorry, our team did not accept to share codes to public! It is a team work, and I can not make it as public.

**Comment by fishfishson** [comment]
@isantesteban hi!
What's the role of trans_vel? is it related to the result of cloth simulation?

**Comment by fishfishson** [comment]
I run run_snug.py with trans_vel 1 and 10 and get totally different results:
![image](https://user-images.githubusercontent.com/28746110/173532790-ff1fc343-cc8c-44b1-af1e-569691ead3b3.png)
It's kind of strange since the output should only depends on smpl shape and pose parameters.

**Comment by fishfishson** [comment]
@isantesteban  It's my fault. The input of the regressor contains trans_vel.

**Comment by fishfishson** [comment]
> @hsyntemiz Hello,
> 
> Have you succeeded in replicating the training. I'm trying to replicate with pytorch, but the loss seems too large. The untrained loss is like: strain:95793.54688, bending:22.87846, gravity:27.13613, inertia:11.21035, collision:18.52707
> 
> Do you have the same observation?

Hi I also try to implement training process with pytorch and find the strain loss is extremly high too
` train/e_stretch=1.24e+5, train/e_bend=24.60, train/e_gravity=0.0449, train/e_inertia=0.0972, train/e_collision=33.80`
Have you solved this problem?

@isantesteban hi, sorry to bother you again. the collision loss in physical.py does not contain k_clossion. Should I multiply this coeff mannually when computing collision loss?

**Comment by Simonhfls** [comment]
> Hi, As you do, I also try to replicate the training part. I have a similar pipeline. But, I couldn't succeed up to now. Do we need postprocessing in training? I am not sure.
> 
> I also have some additional gradient issues. If I don't multiply energy_stretch with a small value, loss terms become NaN immediately. But, these are energy terms. I think it is not correct to add coefficients for energy terms.
> 
> loss_value = energy_bend + g_energy + inertia + 0.01*energy_stretch + penalty_collision
> 
> I am open to collaborate to replicate training part.
> 
> * One sample line from training log for energy values. Strech energy is commonly more than others (~1000x or much more)
>   Bending Energy:0.0080470,Gravitation: 0.058480,Inertia:8.7634e-05, Stretch: 1.305195,Penalty: 9.94163e-05

Hi @hsyntemiz , I encountered the same problem as yours when trying to replicate the training part. Did you manage to solve the issue?
For my case, the results are quite jittery even after 200 epochs.

epoch250, batch50: 
L_inertia=0.07542464137077332,L_strain=1.9437837600708008,L_bend=0.013448243029415607,L_gravity=0.07408734411001205,L_collision=0.009295579977333546

<img width="291" alt="Screenshot 2023-06-19 at 09 03 56" src="https://github.com/isantesteban/snug/assets/46397976/ddee2f75-6438-43d9-b12a-f121fecdb2c3">

Besides, my trained model can only take sequence of length 3 during testing instead of arbitrary length.

**Comment by Simonhfls** [comment]
> > @hsyntemiz Hello,
> > Have you succeeded in replicating the training. I'm trying to replicate with pytorch, but the loss seems too large. The untrained loss is like: strain:95793.54688, bending:22.87846, gravity:27.13613, inertia:11.21035, collision:18.52707
> > Do you have the same observation?
> 
> Hi I also try to implement training process with pytorch and find the strain loss is extremly high too ` train/e_stretch=1.24e+5, train/e_bend=24.60, train/e_gravity=0.0449, train/e_inertia=0.0972, train/e_collision=33.80` Have you solved this problem?
> 
> @isantesteban hi, sorry to bother you again. the collision loss in physical.py does not contain k_clossion. Should I multiply this coeff mannually when computing collision loss?

Hi @fishfishson, I got the same problem as yours but in TensorFlow. Have you succeeded in implementing the training process?
Thanks

**Comment by Simonhfls** [comment]
> > Hi,
> > We don't use the postprocess during training, only the collision penalty.
> > Here are the values of the energies during the training process, we train for 10 epochs using a learning rate of 0.001 and then we resume the training for 100 additional epochs using a learning rate of 0.0001:
> > Untrained:
> > > **full loss:** 392.6,  **bending**: 1.273, **collisions**: 8.24e-4, **gravity**: 0.053, **inertia**: 0.046, **stretch**: 391
> > 
> > 
> > After 10 epochs (LR 0.001):
> > > **full loss:** 0.782,  **bending**: 0.025, **collisions**: 1.8e-4, **gravity**: 0.054, **inertia**: 0.043, **stretch**: 0.62
> > 
> > 
> > After 100 more epochs (LR 0.0001):
> > > **full loss:** 0.256,  **bending**: 0.011, **collisions**: 9.1e-5, **gravity**: 0.033, **inertia**: 0.031, **stretch**: 0.16
> > 
> > 
> > Hope this helps!
> 
> After 10 epochs (LR 0.001)
> 
> **full loss**: 1.22266, **inertia**: 0.01243, **stretch**: 1.04558, **bending**: 0.03306, **gravity**: 0.04731, **collisions**: 0.08428
> 
> After 100 epochs (LR 0.0001)
> 
> **full loss**: 0.20482, **inertia**: 0.01269, **stretch**: 0.12848, **bending**: 0.01619, **gravity**: 0.03859, **collisions**: 0.00887
> 
> Result: ![text2photo-fae1a4c9990dc3ba4a1d0823b6ce2814](https://user-images.githubusercontent.com/74002741/172907730-026104d7-b73b-4b2e-8c5b-cb4a4272c56b.jpeg)
> 
> Loss functions is looking as correct but our result is not perfect, so can you help me for improve it with your offers & analysis? Thanks for your attention!!!

Hi, @opendeeple , I got a similar result as yours, did you find a way to improve the results?

**Comment by zhiqiang-git** [comment]
> > @hsyntemiz Hello,
> > Have you succeeded in replicating the training. I'm trying to replicate with pytorch, but the loss seems too large. The untrained loss is like: strain:95793.54688, bending:22.87846, gravity:27.13613, inertia:11.21035, collision:18.52707
> > Do you have the same observation?
> 
> Hi I also try to implement training process with pytorch and find the strain loss is extremly high too ` train/e_stretch=1.24e+5, train/e_bend=24.60, train/e_gravity=0.0449, train/e_inertia=0.0972, train/e_collision=33.80` Have you solved this problem?
> 
> @isantesteban hi, sorry to bother you again. the collision loss in physical.py does not contain k_clossion. Should I multiply this coeff mannually when computing collision loss?

Hi, @fishfishson , sorry to bother you, but when i'm trying to replicate this work with pytorch, I found that my loss terms are exactly the same as yours. I am wondering that have you found any reasons or solutions about the extremely high stretch loss? Thanks very much.
PS : stretch loss in the picture is multiplied with 1e-3, so it looks different with yours.
![微信截图_20240130183549](https://github.com/isantesteban/snug/assets/127314627/86a4092d-a756-433c-acc9-c65e203041b2)

**Comment by zhiqiang-git** [comment]
> > > @hsyntemiz Hello,
> > > Have you succeeded in replicating the training. I'm trying to replicate with pytorch, but the loss seems too large. The untrained loss is like: strain:95793.54688, bending:22.87846, gravity:27.13613, inertia:11.21035, collision:18.52707
> > > Do you have the same observation?
> > 
> > 
> > Hi I also try to implement training process with pytorch and find the strain loss is extremly high too ` train/e_stretch=1.24e+5, train/e_bend=24.60, train/e_gravity=0.0449, train/e_inertia=0.0972, train/e_collision=33.80` Have you solved this problem?
> > @isantesteban hi, sorry to bother you again. the collision loss in physical.py does not contain k_clossion. Should I multiply this coeff mannually when computing collision loss?
> 
> Hi, @fishfishson , sorry to bother you, but when i'm trying to replicate this work with pytorch, I found that my loss terms are exactly the same as yours. I am wondering that have you found any reasons or solutions about the extremely high stretch loss? Thanks very much. PS : stretch loss in the picture is multiplied with 1e-3, so it looks different with yours. ![微信截图_20240130183549](https://private-user-images.githubusercontent.com/127314627/300774819-86a4092d-a756-433c-acc9-c65e203041b2.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDY4NDg4MzEsIm5iZiI6MTcwNjg0ODUzMSwicGF0aCI6Ii8xMjczMTQ2MjcvMzAwNzc0ODE5LTg2YTQwOTJkLWE3NTYtNDMzYy1hY2M5LWM2NWUyMDMwNDFiMi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwMjAyJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDIwMlQwNDM1MzFaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0xNDA3ODIzNGM1OTQxMWVhMjk4MDRjMDM4MzBkNTM0MWZkM2E5NTBlM2M4Y2NmNGYyNjY5YjUwYTY5ZjNlYWIwJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.-WHyjGMzUgiS9_687ujMHbGyLG_TT5PHUvwGFibOauk)

Sorry, it's my fault. Looks like an Adam optimizer will handle this problem.
